File: web/src/emulator/movie.ts
// Parser for the .nesmovie format produced by the headless recorder
// (tools/record_demo, python/nesenv). A movie is fully self-contained and
// deterministic: it carries the ROM plus one controller byte per frame, so the
// browser can replay exactly what the agent did, frame for frame.
//
// Layout (little-endian):
//   0   "NESMOVIE" (8 bytes magic)
//   8   u32 version (currently 1)
//   12  u32 rom_len
//   16  u32 frame_count
//   20  rom bytes (rom_len)
//   20+rom_len  input bytes (frame_count, one P1 button mask per frame)

export interface NesMovie {
  version: number;
  rom: Uint8Array;
  inputs: Uint8Array; // one byte per frame, player-1 button bitmask
}

const MAGIC = "NESMOVIE";

export function parseMovie(bytes: Uint8Array): NesMovie {
  if (bytes.length < 20) throw new Error("movie too small");
  const magic = new TextDecoder("ascii").decode(bytes.subarray(0, 8));
  if (magic !== MAGIC) throw new Error("not a .nesmovie file");

  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const version = view.getUint32(8, true);
  const romLen = view.getUint32(12, true);
  const frameCount = view.getUint32(16, true);

  const romStart = 20;
  const inputStart = romStart + romLen;
  if (inputStart + frameCount > bytes.length) {
    throw new Error("movie is truncated");
  }

  return {
    version,
    rom: bytes.subarray(romStart, inputStart),
    inputs: bytes.subarray(inputStart, inputStart + frameCount),
  };
}

File: python/nesenv/movie.py
"""Read and write the .nesmovie format that NES Studio's "Watch Movie" replays.

A movie is self-contained: the ROM plus one controller byte per frame. Because the
core is deterministic, replaying it (natively or in the browser) reproduces exactly
what produced it.

Layout (little-endian):
    0   b"NESMOVIE"
    8   uint32 version (1)
    12  uint32 rom_len
    16  uint32 frame_count
    20  rom bytes
    20+rom_len  input bytes (one player-1 mask per frame)
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from pathlib import Path

_MAGIC = b"NESMOVIE"


@dataclass
class Movie:
    version: int
    rom: bytes
    inputs: bytes


def write_movie(path: str | Path, rom: bytes, inputs: bytes | bytearray | list[int]) -> None:
    data = bytes(inputs)
    with open(path, "wb") as f:
        f.write(_MAGIC)
        f.write(struct.pack("<III", 1, len(rom), len(data)))
        f.write(rom)
        f.write(data)


def read_movie(path: str | Path) -> Movie:
    with open(path, "rb") as f:
        blob = f.read()
    if len(blob) < 20 or blob[:8] != _MAGIC:
        raise ValueError("not a .nesmovie file")
    version, rom_len, frames = struct.unpack_from("<III", blob, 8)
    rom = blob[20 : 20 + rom_len]
    inputs = blob[20 + rom_len : 20 + rom_len + frames]
    if len(rom) != rom_len or len(inputs) != frames:
        raise ValueError("movie is truncated")
    return Movie(version=version, rom=rom, inputs=inputs)
